### ISSUE TITLE
Building 16|17 fails on macOS 10.13 due to incompatible 'optional' header

### ISSUE THREAD

**Issue by ilg-ul** [llvm/llvm-project#64472]
While trying to build 16.0.6 on an older macOS, I got this:

```console
$ /usr/bin/clang++ -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/x86_64-apple-darwin17.7.0/build/llvm-16.0.6/lib/Support -I/Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/sources/llvm-project-16.0.6.src/llvm/lib/Support -I/Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/x86_64-apple-darwin17.7.0/build/llvm-16.0.6/include -I/Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/sources/llvm-project-16.0.6.src/llvm/include -I/Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/x86_64-apple-darwin17.7.0/install/include  -ffunction-sections -fdata-sections -pipe -O2  -w -stdlib=libc++ -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -w -fdiagnostics-color  -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk -mmacosx-version-min=10.13 -std=c++17 -MD -MT lib/Support/CMakeFiles/LLVMSupport.dir/Program.cpp.o -MF lib/Support/CMakeFiles/LLVMSupport.dir/Program.cpp.o.d -o lib/Support/CMakeFiles/LLVMSupport.dir/Program.cpp.o -c /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/sources/llvm-project-16.0.6.src/llvm/lib/Support/Program.cpp -v


Apple LLVM version 10.0.0 (clang-1000.10.44.4)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
 "/Library/Developer/CommandLineTools/usr/bin/clang" -cc1 -triple x86_64-apple-macosx10.13.0 -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -emit-obj -disable-free -disable-llvm-verifier -discard-value-names -main-file-name Program.cpp -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -fno-strict-return -masm-verbose -munwind-tables -target-cpu penryn -dwarf-column-info -debugger-tuning=lldb -target-linker-version 409.12 -v -ffunction-sections -fdata-sections -coverage-notes-file /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/x86_64-apple-darwin17.7.0/build/llvm-16.0.6/lib/Support/CMakeFiles/LLVMSupport.dir/Program.cpp.gcno -resource-dir /Library/Developer/CommandLineTools/usr/lib/clang/10.0.0 -dependency-file lib/Support/CMakeFiles/LLVMSupport.dir/Program.cpp.o.d -skip-unused-modulemap-deps -sys-header-deps -MT lib/Support/CMakeFiles/LLVMSupport.dir/Program.cpp.o -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk -D __STDC_CONSTANT_MACROS -D __STDC_FORMAT_MACROS -D __STDC_LIMIT_MACROS -I /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/x86_64-apple-darwin17.7.0/build/llvm-16.0.6/lib/Support -I /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/sources/llvm-project-16.0.6.src/llvm/lib/Support -I /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/x86_64-apple-darwin17.7.0/build/llvm-16.0.6/include -I /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/sources/llvm-project-16.0.6.src/llvm/include -I /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/x86_64-apple-darwin17.7.0/install/include -D NDEBUG -stdlib=libc++ -O3 -Werror=date-time -Werror=unguarded-availability-new -w -std=c++17 -fdeprecated-macro -fdebug-compilation-dir /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/x86_64-apple-darwin17.7.0/build/llvm-16.0.6 -ferror-limit 19 -fmessage-length 0 -fvisibility-inlines-hidden -stack-protector 1 -fblocks -fencode-extended-block-signature -fobjc-runtime=macosx-10.13.0 -fcxx-exceptions -fexceptions -fmax-type-align=16 -fdiagnostics-show-option -fcolor-diagnostics -vectorize-loops -vectorize-slp -o lib/Support/CMakeFiles/LLVMSupport.dir/Program.cpp.o -x c++ /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/sources/llvm-project-16.0.6.src/llvm/lib/Support/Program.cpp
clang -cc1 version 10.0.0 (clang-1000.10.44.4) default target x86_64-apple-darwin17.7.0
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/c++/v1"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/x86_64-apple-darwin17.7.0/build/llvm-16.0.6/lib/Support
 /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/sources/llvm-project-16.0.6.src/llvm/lib/Support
 /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/x86_64-apple-darwin17.7.0/build/llvm-16.0.6/include
 /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/sources/llvm-project-16.0.6.src/llvm/include
 /Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/x86_64-apple-darwin17.7.0/install/include
 /Library/Developer/CommandLineTools/usr/include/c++/v1
 /Library/Developer/CommandLineTools/usr/lib/clang/10.0.0/include
 /Library/Developer/CommandLineTools/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/System/Library/Frameworks (framework directory)
End of search list.
/Users/ilg/Work/xpack-dev-tools/clang-xpack.git/build/darwin-x64/sources/llvm-project-16.0.6.src/llvm/lib/Support/Program.cpp:46:49: error: no viable constructor or deduction guide for deduction of template arguments of 'optional'
        PI, SecondsToWait == 0 ? std::nullopt : std::optional(SecondsToWait),
                                                ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:684:41: note: candidate template ignored: couldn't infer template argument '_Tp'
    _LIBCPP_INLINE_VISIBILITY constexpr optional(nullopt_t) noexcept {}
                                        ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:690:24: note: candidate template ignored: couldn't infer template argument '_Tp'
    constexpr explicit optional(in_place_t, _Args&&... __args)
                       ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:704:15: note: candidate template ignored: couldn't infer template argument '_Tp'
    constexpr optional(_Up&& __v)
              ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:711:24: note: candidate template ignored: couldn't infer template argument '_Tp'
    constexpr explicit optional(_Up&& __v)
                       ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:580:7: note: candidate template ignored: could not match 'optional<_Tp>' against 'unsigned int'
class optional
      ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:682:41: note: candidate template ignored: could not match 'optional<_Tp>' against 'unsigned int'
    _LIBCPP_INLINE_VISIBILITY constexpr optional(const optional&) = default;
                                        ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:683:41: note: candidate template ignored: could not match 'optional<_Tp>' against 'unsigned int'
    _LIBCPP_INLINE_VISIBILITY constexpr optional(optional&&) = default;
                                        ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:719:5: note: candidate template ignored: could not match 'optional<type-parameter-0-1>' against 'unsigned int'
    optional(const optional<_Up>& __v)
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:727:14: note: candidate template ignored: could not match 'optional<type-parameter-0-1>' against 'unsigned int'
    explicit optional(const optional<_Up>& __v)
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:737:5: note: candidate template ignored: could not match 'optional<type-parameter-0-1>' against 'unsigned int'
    optional(optional<_Up>&& __v)
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:745:14: note: candidate template ignored: could not match 'optional<type-parameter-0-1>' against 'unsigned int'
    explicit optional(optional<_Up>&& __v)
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:681:41: note: candidate function template not viable: requires 0 arguments, but 1 was provided
    _LIBCPP_INLINE_VISIBILITY constexpr optional() noexcept {}
                                        ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:697:24: note: candidate function template not viable: requires at least 2 arguments, but 1 was provided
    constexpr explicit optional(in_place_t, initializer_list<_Up> __il, _Args&&... __args)
                       ^
1 error generated.
```

Apparently some new changes introduced in 16.x require a newer version of the `optional` header than the one provided by clang 10.

The same build scripts pass on Intel macOS 12, and Apple Silicon macOS 11.

I know that macOS 10.13 is quite old, but it is more or less the de-facto build platform for standalone binary distributions for Intel macOS, including the [xPack 3rd Party Development Tools](https://github.com/xpack-dev-tools/).

To be totally fair, for me this was not a show stopper, since in my build environment I have the luxury to easily switch between tools versions, and so I compiled 16.0.6 using 15.0.7, even on the old macOS 10.13, but sometimes I still need to calibrate my builds using the system toolchains, and having the LLVM source files compatible with macOS 10.13 would be a useful feature.

**Comment by Endilll** [comment]
Do you still experience this with Clang 17?

**Comment by ilg-ul** [comment]
I did not try it yet, I'll report back when I'll have more results.

**Comment by ilg-ul** [comment]
I tried a build on 10.14.6, and with clang 17 there are **a lot** of messages, apparently more than with clang 16.

Attached is the build console log.

[build-output-20230928-131857.txt.zip](https://github.com/llvm/llvm-project/files/12749884/build-output-20230928-131857.txt.zip)

**Comment by ldionne** [comment]
Removing the libc++ tag since this is not a libc++ issue. The OP is trying to compile LLVM/Clang against an older macOS SDK (`MacOSX10.14.sdk` according to the logs) and using `Apple LLVM version 10.0.0`. Whether LLVM and Clang want to support that is up to them, not libc++. There's also nothing for libc++ to fix in this case because this issue in `<optional>` has been fixed in trunk since long.

**Comment by Endilll** [comment]
I wasn't sure how to tag this. Thank you for taking a look!

**Comment by ldionne** [comment]
> I wasn't sure how to tag this. Thank you for taking a look!

No worries, and thanks for tagging issues in the first place :-) It's really helpful to have people screening incoming issues, and some amount of re-routing is both fine and unavoidable.

**Comment by ulion** [comment]
@ilg-ul I'm curious, where is the c++ header files came from of the llvm@15, I built it on osx 10.13, with native tool-chain, but result no c++ headers in /usr/local/opt/llvm@15 folder, neither llvm@12, llvm@14. so I also failed to build llvm@16 with llvm@15, since it do not have any c++ header files, I have to add -I flag to use system c++ headers from xcode/cmd tools, and they result the same std::optional issue you posted here (with llvm@15, not system clang-10).
So, I must miss something, which cause I do not have some c++/v1 folder in any llvm build result. do you have those c++ header files in your llvm@15 folder? any idea why they are missing and where are they from? (they are not in llvm source neither)
If you can build llvm@16 with llvm@15, then you must have the updated std::optional header file somewhere? where did they come from then? thx.

**Comment by ilg-ul** [comment]
@ulion, I'm afraid I don't fully understand your question. The issue is pretty old and I can no longer remember details. Now I'm facing other issues with clang 17... :-(

**Comment by ulion** [comment]
thx, I think my llvm old version build are missing libc++, it was not in anywhere after brew build.

**Comment by ldionne** [comment]
As @ilg-ul said above, this issue doesn't seem relevant. If you are encountering a broken installation via Homebrew, I suggest you begin by filing an issue against Homebrew, and we can take it back to LLVM if it's truly a LLVM issue. However, this is not super likely because the world would have noticed if we stopped shipping the `<optional>` header any time.

Closing to keep things tidy, please reopen if you think this still has value.

**Comment by ilg-ul** [comment]
@ldione, I'm afraid you did not understand the issue. :-(

The issue has nothing to do with Homebrew or any other distribution.

The issue occurs when trying to build clang >= 16 on an older macOS, like 10.13, using the Apple compiler.

It is easy to reproduce, you login to an old macOS 10.13, install latest CLT available, clone the llvm repo and invoke that single command presented in the initial message. It fails. Not good.

You can run the full build but the result will be the same, when you get to that file, the build will fail.

The reason it fails is that starting from clang 16, the clang sources assume some content of the headers, not available in macOS 10.13. Not nice.

**Comment by ldionne** [comment]
Ah, I see. Thanks for breaking it down.

What is the Clang support policy for older compilers? According to https://llvm.org/docs/GettingStarted.html#host-c-toolchain-both-compiler-and-standard-library, the oldest supported Apple Clang seems to be Apple Clang 10. But when you compiled with Apple Clang 10 [here](https://github.com/llvm/llvm-project/issues/64472#issuecomment-1739176660), you reported a bunch of errors.

So basically, it looks like LLVM/Clang is not upholding its stated compiler support policy of AppleClang 10 (and the same thing for older non-Apple Clang versions too). @AaronBallman any thoughts on this? Is that intentional or just the product of a lack of testing on older compilers?

**Comment by AaronBallman** [comment]
> Ah, I see. Thanks for breaking it down.
> 
> What is the Clang support policy for older compilers? According to https://llvm.org/docs/GettingStarted.html#host-c-toolchain-both-compiler-and-standard-library, the oldest supported Apple Clang seems to be Apple Clang 10. But when you compiled with Apple Clang 10 [here](https://github.com/llvm/llvm-project/issues/64472#issuecomment-1739176660), you reported a bunch of errors.
> 
> So basically, it looks like LLVM/Clang is not upholding its stated compiler support policy of AppleClang 10 (and the same thing for older non-Apple Clang versions too). @AaronBallman any thoughts on this? Is that intentional or just the product of a lack of testing on older compilers?

This is a product of lack of testing with older compilers rather than anything intentional, as best I can tell. Personally, I think that we should have a community requirement that there is post-commit CI doing a bootstrap build of Clang with each of the oldest compilers we claim to support. If we don't have CI building it (or a volunteer actively testing it in some downstream and reporting issues back to the community), we shouldn't be claiming we support it. I don't think we have anything in the lab building with Clang 5.0 either, at least from what I can tell from a cursory look.

That said, the idea of requiring CI to test our lowest common denominator has been raised in the past but didn't go anywhere; we're ultimately at the mercy of whoever volunteers to host builders in the lab. CC @llvm-beanz @rnk for ideas on what we might be able to reasonably accomplish here.

**Comment by ldionne** [comment]
FWIW I strongly agree with this. If something matters and we want to guarantee it, we should have a test on it.

If we follow this principle strictly and only claim to support what we test, we generally find that we actually support a much narrower set of compilers than we might think. For libc++ we pinned that down by being quite strict about our compiler requirements, but testing everything we claim to support in our CI. This has worked really well but clearly we have different requirements and incentives to support older compilers than the Clang codebase does itself.

**Comment by ilg-ul** [comment]
> On 12 Apr 2024, at 17:03, Aaron Ballman ***@***.***> wrote:
> 
> This is a product of lack of testing with older compilers ...

Ok, that was my guess too, and this is why I raised this issue.

> I think that we should have a community requirement that there is post-commit CI doing a bootstrap build of Clang with each of the oldest compilers we claim to support.

Fully agree.

> ... we're ultimately at the mercy of whoever volunteers to host builders in the lab. CC @llvm-beanz @rnk for ideas on what we might be able to reasonably accomplish here.

Travis CI has a wide range of older macOS machines available.

- https://docs.travis-ci.com/user/reference/osx/#macos-version

I know that they are a commercial company, but they continue to support open source project, for example they allow me to run my tests on their machines for free.

You can try to contact them, perhaps they are reasonable. How long would a build test take?

 
I also have an Intel macOS 10.13 machine that runs self-hosted GitHub runners for my production builds.

If that helps I can run the llvm build tests on it too, just that the machine is infrequently used and most of the time is not powered up; but if you have a notification mechanism for me to power up the machine and run the test, this might work, at least temporarily.


Liviu

**Comment by AaronBallman** [comment]
Thank you for the suggestion -- CC @tstellar @tru @joker-eph for awareness

**Comment by tru** [comment]
Hmmm. I doubt we would like to add more CI vendors at this point. It's hard enough to get people to maintain the current github + buildkite configuration. But yes - support for "minimum" platforms are a hard problem that should be supported better, not sure how to manage that with the current volunteer maintainers though.

**Comment by ilg-ul** [comment]
It looks like it'll take some time to find a solution for the CI...

Till then, could someone take a look at the actual problem reported here and find a solution (probably some #ifdefs) to make the code compilable on macOS 10.13?

**Comment by PHHargrove** [comment]
@ilg-ul 
In my own testing just now with macOS High-Sierra, the following trivial change resolved the error you reported.
No `#ifdef`s required.
Can you confirm this works for you as well?

```diff
diff --git a/llvm/lib/Support/Program.cpp b/llvm/lib/Support/Program.cpp
index 181f68c..47f24f8 100644
--- a/llvm/lib/Support/Program.cpp
+++ b/llvm/lib/Support/Program.cpp
@@ -43,7 +43,7 @@ int sys::ExecuteAndWait(StringRef Program, ArrayRef<StringRef> Args,
     if (ExecutionFailed)
       *ExecutionFailed = false;
     ProcessInfo Result = Wait(
-        PI, SecondsToWait == 0 ? std::nullopt : std::optional(SecondsToWait),
+        PI, SecondsToWait == 0 ? std::nullopt : std::optional<unsigned>(SecondsToWait),
         ErrMsg, ProcStat);
     return Result.ReturnCode;
   }
```

**Comment by ilg-ul** [comment]
> Can you confirm this works for you as well?

Unfortunately I no longer have the 10.13 build machine around, I upgraded it to 11.7. :-(

**Comment by hfath** [comment]
Came here after googling the error message... I am trying to build lang/llvm (18.1.8) from pkgsrc on macOS 10.13 "High Sierra", the last release supported on this hardware.

The above patch fixes the first build failure for me, but there are more:
```
[hauke@pizza] /<5>work/llvm-18.1.8.src > fgrep -R 'std::optional(' * 
lib/Support/Program.cpp:        PI, SecondsToWait == 0 ? std::nullopt : std::optional(SecondsToWait),
lib/Transforms/InstCombine/InstructionCombining.cpp:  return std::optional(std::pair(L0, R0));
lib/Transforms/Scalar/LowerMatrixIntrinsics.cpp:            IsIntVec ? std::nullopt : std::optional(FMF)) +
lib/Transforms/Utils/MemoryOpRemark.cpp:      TySize ? std::optional(TySize->getFixedValue()) : std::nullopt;
[hauke@pizza] /<5>work/llvm-18.1.8.src >
```
From my rusty C++ fu: What is the significance of std::optional(_expression_), as opposed to std::optional<T>(_expression_)? I haven't been able to find anything helpful on the web.

In particular, I have a hard time finding the <T> equivalent to the above _std::pair(L0, R0)_.
